---
title: "WorkflowHub: a registry for computational workflows"
categories:
  - Publication
summary: >
  arXiv preprint 
description: >
  Computational workflows are a special kind of software for handling complex data pipelines, analyses and simulations. Their rising popularity is driven by the need for repetitive and scalable data processing, access to and exchange of processing know-how, and the desire for more reproducible/ transparent and quality assured processing methods. As records of analyses, they should be reviewable and available alongside their derived data. As descriptions of software processing steps, they should support reproducibility, and be reused and adapted. Therefore, similar to data and other software tools, workflows should be shared and their authors should be properly recognized given the time, effort, and expertise required to create them. 
  Sharing workflows presents the opportunity for reducing unnecessary reinvention, enhancing knowledge sharing, promoting reuse and productivity, and democratising access to computational processes that can be beyond the skills of many researchers. However, in practice, workflows are scattered across repositories, in part due to the different workflow engines they are written for and the workflow ecosystems where they were born, but also because workflow sharing is not yet a central part of research practice. 
  WorkflowHub provides a FAIR, unified registry of workflows with rich metadata support that links with community repositories and supports the development and publishing lifecycle of workflows. It aims to support making workflows findable, accessible, interoperable, and reusable (FAIR), particularly emphasising machine- and human-readable metadata to enhance FAIRness, while explicitly addressing practical concerns of assigning credit appropriately to individuals and consortia. By working and interoperating with different workflow platforms, services, and external registries, WorkflowHub provides added value capabilities to support workflow providers and users, and promote workflows as scholarly artefacts.
  WorkflowHub serves as a Community of Practice Hub, supporting workflow subscriptions and community channels. It is agnostic to the workflow manager used, accommodating any workflow types, for any discipline at any stage of the workflow development cycle. WorkflowHub has a global reach with hundreds of research organisations and projects, including major European and Australian organisations like ELIXIR and Australian BioCommons. Indeed, over its first four years, over 250 Teams and more than 700 workflows have been registered.
tags:
  - FAIR
  - Workflow  
authors:
  - name: Ove Johan Ragnar Gustafsson
    orcid: https://orcid.org/0000-0002-2977-5032
  - name: Sean Wilkinson
    orcid: https://orcid.org/0000-0002-1443-7479
  - name: Finn Bacall
    orcid: https://orcid.org/0000-0002-0048-3300
  - name: Stian Soiland-Reyes
    orcid: https://orcid.org/0000-0001-9842-9718
  - name: Simone Leo
    orcid: https://orcid.org/0000-0001-8271-5429
  - name: Luca Pireddu
    orcid: https://orcid.org/0000-0002-4663-5613
  - name: Stuart Owen
    orcid: https://orcid.org/0000-0003-2130-0865
  - name: Nick Juty
    orcid: https://orcid.org/0000-0002-2036-8350
  - name: José Mª Fernández
    orcid: https://orcid.org/0000-0002-4806-5140    
  - name: Tom Brown
    orcid: https://orcid.org/0000-0001-8293-4816
  - name: Hervé Ménager
    orcid: https://orcid.org/0000-0002-7552-1009
  - name: Björn Grüning
    orcid: https://orcid.org/0000-0002-3079-6586
  - name: Salvador Capella-Gutierrez
    orcid: https://orcid.org/0000-0002-0309-604X
  - name: Frederik Coppens
    orcid: https://orcid.org/0000-0001-6565-5145
  - name: WorkflowHub community
    orcid: https://about.workflowhub.eu/project/acknowledgements/#workflowhub-club
  - name: Carole Goble
    orcid: https://orcid.org/0000-0003-1219-2137
---

**Note**: This preprint has not yet been fully adapted for the Web.

<h2>Cite as</h2>

Ove Johan Ragnar Gustafsson, Sean R. Wilkinson, Finn Bacall, Stian Soiland-Reyes, Simone Leo, Luca Pireddu, Stuart Owen, Nick Juty, José Mª Fernández, Tom Brown, Hervé Ménager, Björn Grüning, Salvador Capella-Gutierrez, Frederik Coppens, WorkflowHub community, Carole Goble (2024):  
**WorkflowHub: a registry for computational workflows**.  
(in preparation)

# WorkflowHub: a registry for computational workflows

_Ove Johan Ragnar Gustafsson<sup>1</sup>, Sean R. Wilkinson<sup>2</sup>, Finn Bacall<sup>3</sup>, Stian Soiland-Reyes<sup>3</sup>, Simone Leo<sup>4</sup>, Luca Pireddu<sup>4</sup>, Stuart Owen<sup>3</sup>, Nick Juty<sup>3</sup>, José Mª Fernández<sup>5,6</sup>, Tom Brown<sup>7</sup>, Hervé Ménager<sup>8,9</sup>, Björn Grüning<sup>10</sup>, Salvador Capella-Gutierrez<sup>6,11</sup>, Frederik Coppens<sup>12</sup>, WorkflowHub community, Carole Goble<sup>3</sup>_


<div class="affiliations">

<sup>1</sup> Australian BioCommons, University of Melbourne, Melbourne, Victoria, Australia   
<sup>2</sup> Oak Ridge National Laboratory, Oak Ridge, TN, USA   
<sup>3</sup> The University of Manchester, Manchester, UK   
<sup>4</sup> Center for Advanced Studies, Research, and Development in Sardinia (CRS4), Pula (CA), Italy  
<sup>5</sup> Barcelona Supercomputing Center (BSC), Spain  
<sup>6</sup> Spanish National Bioinformatics Institute (INB), Spain  
<sup>7</sup> Leibniz Institute for Zoo- and Wildlife Research, Berlin, Germany  
<sup>8</sup> Institut Pasteur, Université Paris Cité, Bioinformatics and Biostatistics Hub, 75015, Paris, France  
<sup>9</sup> CNRS, UMS 3601, Institut Français de Bioinformatique, IFB-core, 2 rue Gaston Crémieux, F-91000 Evry, France  
<sup>10</sup> Albert-Ludwigs-Universität Freiburg, Freiburg, Germany  
<sup>11</sup> Barcelona Supercomputing Center (BSC), Spain  
<sup>12</sup> VIB Data Core, VIB Technologies, Ghent, Belgium

</div>

* **License**: Creative Commons Attribution License ([CC BY 4.0](https://spdx.org/licenses/CC-BY-4.0)). 
* **Modifications**: Reformatting as Markdown; references in s11 house style; inline citation hyperlinks; inline hyperlinks. 



## Abstract

Computational workflows are a special kind of software for handling complex data pipelines, analyses and simulations. Their rising popularity is driven by the need for repetitive and scalable data processing, access to and exchange of processing know-how, and the desire for more reproducible/ transparent and quality assured processing methods. As records of analyses, they should be reviewable and available alongside their derived data. As descriptions of software processing steps, they should support reproducibility, and be reused and adapted. Therefore, similar to data and other software tools, workflows should be shared and their authors should be properly recognized given the time, effort, and expertise required to create them. 

Sharing workflows presents the opportunity for reducing unnecessary reinvention, enhancing knowledge sharing, promoting reuse and productivity, and democratising access to computational processes that can be beyond the skills of many researchers. However, in practice, workflows are scattered across repositories, in part due to the different workflow engines they are written for and the workflow ecosystems where they were born, but also because workflow sharing is not yet a central part of research practice. 

WorkflowHub provides a FAIR, unified registry of workflows with rich metadata support that links with community repositories and supports the development and publishing lifecycle of workflows. It aims to support making workflows findable, accessible, interoperable, and reusable (FAIR), particularly emphasising machine- and human-readable metadata to enhance FAIRness, while explicitly addressing practical concerns of assigning credit appropriately to individuals and consortia. By working and interoperating with different workflow platforms, services, and external registries, WorkflowHub provides added value capabilities to support workflow providers and users, and promote workflows as scholarly artefacts.

WorkflowHub serves as a Community of Practice Hub, supporting workflow subscriptions and community channels. It is agnostic to the workflow manager used, accommodating any workflow types, for any discipline at any stage of the workflow development cycle. WorkflowHub has a global reach with hundreds of research organisations and projects, including major European and Australian organisations like ELIXIR and Australian BioCommons. Indeed, over its first four years, over 250 Teams and more than 700 workflows have been registered.  

## Introduction

In an era of Big Data and data-driven science, the need for repetitive, scalable, reproducible and quality-assured data processing and analysis methods has contributed to a surge in popularity for computational workflows. The past two decades have seen a handful of workflow management systems (WfMS) expand to hundreds \[[**Amstutz2024**](https://s.apache.org/existing-workflow-systems)\], and workflows applied across a growing number of domains, including biosciences \[[**Maier et al. 2021**](https://doi.org/10.1038/s41587-021-01069-1)\], astronomy \[[**Freudling et al 2023**](https://doi.org/10.48550/arXiv.2311.03822)\] and the physical sciences \[[**McClure et al 2020**](https://doi.org/10.1007/978-3-030-63393-6_15)\].

In brief, computational workflows are a special kind of software for handling multi-step, multi-code data pipelines, analysis, and simulations, and are intended to automate data-handling processes. They come in many forms, but typically share certain features: a high-level language executed by a dedicated WfMS, which manages data flow and code execution; a composition of modular code or workflow building blocks that can be remixed; and a tendency to be closely associated (i.e. intertwined) with the data on which they will operate \[[**Gil2007**](https://doi.org/10.1109/MC.2007.421)\]. Important scientific goals like repeatability, replicability, and reproducibility become more realistic when scientists specify their experiment’s analysis processes as a computational workflow \[[**Cohen-Boulakia2017**](https://doi.org/10.1016/j.future.2017.01.012)\]. For example, computational workflows have become central to major international science missions that require systematic, reproducible, and shared data analysis. Recent examples include the global response to the COVID-19 pandemic and the analyses of SARS-CoV-2 \[[**Maier et al. 2021**](https://doi.org/10.1038/s41587-021-01069-1)\], and the large-scale sequencing efforts currently in-flight for the Vertebrate Genomes Project (VGP) \[[**Larivière et al. 2024**](https://doi.org/10.1038/s41587-023-02100-3)\]. While these large consortia with defined collaborative research programs are key drivers for computational workflow creation and deployment, workflows are also being adopted across scientific disciplines as their requirements become increasingly computational, and the emphasis on reproducibility and portability increases \[[**ICGC2020**](https://doi.org/10.1038/s41586-020-1969-6)\]\[[**Reiter2021**](https://doi.org/10.1093/gigascience/giaa140)\]\[[**Wratten2021**](https://doi.org/10.1038/s41592-021-01254-9)\]\[[**Patel2022**](https://doi.org/10.48550/arXiv.2205.10267)\].

Increasingly, scientists are also being asked to share their data and associated research objects (i.e. software), in ways others can reuse \[[**NelsonMemo2022**](https://www.whitehouse.gov/wp-content/uploads/2022/08/08-2022-OSTP-Public-Access-Memo.pdf)**, [NasaSPD41a](https://smd-cms.nasa.gov/wp-content/uploads/2023/08/smd-information-policy-spd-41a.pdf)**\]. The idea is to accelerate scientific progress and spur innovation by enabling scientists to avoid reinventing each others’ work, and to explicitly support confidence in published results by removing ambiguity surrounding the approach taken to create research outcomes. In addition, scientific activity often includes the exploration of analysis variance; modifying workflows to understand effects and changes on data products is simpler when those workflows are clearly described and available. To this end, Wilkinson *et al.* published guiding principles for scientific data management and stewardship, providing guidelines for making data and other research objects Findable, Accessible, Interoperable, and Reusable (FAIR) by others \[[**Wilkinson2016**\]](https://doi.org/10.1038/sdata.2016.18). The FAIR principles have sparked an entire movement in the international community towards adopting FAIR practices, and further work has been undertaken to extend the principles to research software \[[**Barker et al 2022**](https://www.nature.com/articles/s41597-022-01710-x)\], AI models \[[**Huerta et al 2023**](https://www.nature.com/articles/s41597-023-02298-6)\], and computational workflows \[PLACEHOLDER\].. A fundamental step towards supporting FAIR workflows \[[**Goble et al 2020**](https://doi.org/10.1162/dint_a_00033)\] is to enable the sharing of workflows and their descriptions \[[Gil et al 2009](https://doi.org/10.3233/SPR-2009-0261)\] and make them findable.

Researchers find software by searching: the web (i.e. search engines), public software project repositories (e.g. GitHub), the literature, mailing lists, discussion groups (e.g. StackOverflow), dependencies in the software itself, relevant registries (e.g. CRAN, communities of practice like nf-core \[[**Ewels2020**](https://nf-co.re/)\]\[[**Di Tommaso et al. 2017**](https://doi.org/10.1038/nbt.3820)\]), and even social media \[[**Stevens2022**](https://doi.org/10.5281/zenodo.7340034)\]. High-quality machine-processable metadata markup is needed to make workflows more findable and understandable in such a “search context”: in other words, the descriptors for workflows must be themselves standardised, accessible, and discoverable. Current mechanisms for sharing workflows do not achieve this outcome for the entire workflow ecosystem. Sharing source code includes options such as version control platforms (e.g. GitHub [https://github.com/](https://github.com/) or GitLab [https://about.gitlab.com/](https://about.gitlab.com/)), and WfMS-specific curated git repositories (e.g. Intergalactic Workflow Commission [https://github.com/galaxyproject/iwc](https://github.com/galaxyproject/iwc) \[[**GalaxyCommunity2024**](https://doi.org/10.1093/nar/gkae410)\], nf-core, Snakemake catalogue \[[**The Snakemake team 2020**](https://snakemake.github.io/snakemake-workflow-catalog/)\]\[[**Molder et al. 2021**](https://doi.org/10.12688/f1000research.29032.2)\]). Creators can also publish their workflows, either in public generalist repositories (e.g. Zenodo [https://zenodo.org/](https://zenodo.org/), DataVerse [https://dataverse.org/](https://dataverse.org/)), conventional journals (e.g. GigaScience, [https://academic.oup.com/gigascience](https://academic.oup.com/gigascience)) or software journals (e.g. the Journal of Open Source Software, JOSS, [https://joss.theoj.org/](https://joss.theoj.org/)). Finally, a creator can register the workflow using either a platform-specific (e.g. Knime Community Hub https://hub.knime.com/, BinderHub [https://binderhub.readthedocs.io/en/latest/](https://binderhub.readthedocs.io/en/latest/), nf-core [https://nf-co.re/](https://nf-co.re/)) or platform-agnostic solution. The latter includes Dockstore, a registry that supports the sharing and running of containerised tools, workflows and notebooks across diverse cloud computing [environments](https://dockstore.org/) \[[**Yuen2021**](https://doi.org/10.1093/nar/gkab346)\].

While critical for software, discovery of these many resources and platforms can be impeded by non-standardised descriptors that are not necessarily visible to search. Even once a workflow has been found, a divergent ecosystem does not lend itself to better integration of services, adoption of standards, or achieving FAIR outcomes for workflows. A registry that serves as a hub for these various mechanisms, and their specific benefits, would begin to address these challenges. It could support workflow developers to share and gain credit for their work, integrate with the platforms, services and infrastructures that developers and users rely on to both create and use workflows, and support making workflows FAIR. Structurally, a registry should be flexible, extensible, and use internationally recognised standards that accommodate rich metadata. To capture and present the breadth of the global computational workflow ecosystem back to the research community, a registry should be agnostic to domains and WfMS, and embrace community standards. Finally, it should provide mechanisms that can link workflows to other digital objects that provide context for a research project, including documents, standard operating procedures (SOPs) and publications.

Here, we present a public and inclusive registry dedicated specifically to the sharing of computational workflows: [WorkflowHub](https://workflowhub.org) \[[**Goble2021**](https://doi.org/10.5281/zenodo.4605654)\]\[[**Goble2022**](https://doi.org/10.7490/f1000research.1118984.1)\]\[[**Goble2023**](https://doi.org/10.5281/zenodo.7886545)\]. WorkflowHub is designed to allow any scientist, regardless of expertise level, to contribute and share computational workflows. It indexes workflows from any scientific domain, in any format, in any workflow language, regardless of whether it uses a WfMS. Here, we describe in detail how WorkflowHub’s structure, design, standards, community engagement, and continued evolution support: 1\) collaboration, sharing and credit for workflow developers, projects, and consortia; 2\) integration with added-value services, platforms, and capabilities that support the workflow life cycle (i.e. creation, version control, execution, maintenance, reuse and citation); and 3\) wizards and inbuilt features that ease the process of sharing workflows alongside the constellation of associated digital artefacts that give a workflow its scientific context (i.e. documents, SOPs, publications).

## Results

### A registry for computational workflows

The WorkflowHub is a registry for describing, sharing and publishing scientific computational workflows, regardless of their type, where they are located and maintained, or their discipline. On the landing page for WorkflowHub, a new user is presented with a description of the platform and its purpose, the latest workflow additions, what content is discoverable, and how to join the WorkflowHub community. Underpinning WorkflowHub is the implementation of open tools and standards, which are further described herein. Collaborating Teams are supported by registry features that support workflow reuse, and include integration with native workflow repositories, assignment of credit, import and export, and the creation of curated Collections of workflows that are enriched by other digital objects (e.g. publications, SOPs). **Figure 1** provides a conceptual view of the WorkflowHub’s capabilities and its relationship to the workflow development and publishing ecosystem outlined above and discussed later in more depth.

![figure1.png](figure1.png)

**Figure 1\.** WorkflowHub connects to platforms, services, and resources that support a workflow’s life cycle [**Courbebaisse2023**](https://doi.org/10.5281/zenodo.8324828)\]. A researcher initially needs to **Plan & Find**, where they either plan for a particular analysis and find existing workflows (i.e. using a registry), or **Develop** a new workflow. WorkflowHub integrates with Git repositories (e.g. GitHub, GitLab), and Git-supported communities (e.g. nf-core), to support development. A workflow requires **Test & Review** to **Run & Deploy**, and here WorkflowHub connects to support services (e.g. LifeMonitor, bio.tools, Sapporo WES, WfExS) and welcomes diverse workflow platforms that aid deployment (e.g. CWL, Snakemake, Galaxy, Jupyter, Python, BASH, WDL, Nextflow). A creator needs to **Share** a workflow and can benefit from WorkflowHub’s use of citation infrastructures and standards (i.e CITATION.cff, Zenodo, DataCite, DOI and ORCID). In the **Maintain & Learn** stage, maintenance, and also understanding of a workflow by other researchers, becomes critical as it impacts workflow **Reuse & Rework**, where a workflow is either reused, or adapted, by other researchers to suit their requirements. WorkflowHub supports these stages through registration of digital objects that enrich a workflow (e.g. documents, publications, SOPs), the ability to create Collections and workflow citations based on DOIs, and ultimately through the connections created to knowledge graphs. WorkflowHub also enables communities of practice to benefit from all its integrations and connections, ensuring that they can reuse or rework workflows from across the globe. The entire support framework is enabled by the implementation of standards that allow WorkflowHub to interact with the ecosystem and truly act as a “Hub”: EDAM, research object crates (RO-crates), GA4GH APIs, common workflow language (CWL), FAIR signposting, and Bioschemas.

The registry is agnostic to domain, discipline and workflow type, supporting its adoption by a wide spectrum of researchers and other stakeholders. As a result, at the time of writing, the registry already indexes workflows for astronomy, biodiversity, climate modelling, and particle physics (see [https://workflowhub.eu/workflows](https://workflowhub.eu/workflows)), with 764 workflows registered spanning a range of workflow types (see [https://workflowhub.eu/workflow\_classes](https://workflowhub.eu/workflow_classes)). In addition, the registry currently has 840 registered users from 236 organisations across 35 countries.

![figure2.png](figure2.png)

**Figure 2\.** Workflow types registered with WorkflowHub. For an up-to-date list of   
workflows visit [https://workflowhub.eu/workflows](https://workflowhub.eu/workflows).

WorkflowHub was launched in 2020, as part of the EOSC-Life Workflow Collaboratory \[[**Goble2023**](https://doi.org/10.5281/zenodo.7886545)\], to support the registration of workflows required for the response to the COVID pandemic. WorkflowHub now houses [66 COVID-related workflows](https://workflowhub.eu/search?utf8=%E2%9C%93&q=covid#workflows), including those that support the ongoing global analysis of intra-host variation as new samples become available \[[**Maier et al. 2021**](https://doi.org/10.1038/s41587-021-01069-1)\]\[[**Baker et al. 2020**](https://doi.org/10.1371/journal.ppat.1008643)\]. This outcome demonstrates a central ambition of WorkflowHub: to be of practical use in advancing the application of computational workflows in research science by *supporting the needs of the communities that it serves*.

WorkflowHub meets community requirements and supports the workflow life cycle in three key ways. Firstly, the registry provides structures that directly support collaboration, sharing knowledge and distributing credit. Secondly, woven into this structure are multiple integrations with other elements of the global research ecosystem that support the workflow life cycle: creation, development, discovery, reuse, and citation of workflows. Finally, WorkflowHub provides a registration wizard that guides users in leveraging these structures and integrations. This approach is deliberate and will continue to evolve in lock step with the requirements of the community. In the following sections, we will first describe how the data model and metadata framework of WorkflowHub support the registry's core functions \- namely registering, finding, and launching workflows and associated digital data objects. In turn we highlight how the registry design, including the use of wizards to guide best practice, allows it to act as an integrating hub across the workflows ecosystem and to support each stage of the workflow life cycle. Finally, we will describe how the WorkflowHub engages with the workflow community and highlight some key use cases for the registry.

### A data model that reflects real-world collaborations that create workflows

Science is a collaborative enterprise, and infrastructure platforms should reflect this quality to be of practical use in accelerating science missions. Research programs and projects are also intertwined with diverse computational approaches and ways of sharing research outcomes, and these are subject to the same requirements for findability, credit and impact assessment [**\[Gil2007\]**](https://doi.org/10.1109/MC.2007.421). As a result, WorkflowHub is structured to reflect real-world collaboration and assign complex credit well. 

The data model of WorkflowHub provides access to three elements for every registered user: Organisations, Teams, and Spaces. A user can specify one or more Organisations (i.e. affiliations) as part of their user profile. They also must belong to at least one Team, which must also belong to an administrative Space. If existing Teams are not suitable or appropriate, a user can create a new Team. A member of a Team can specify Organisations for each Team they join, which allows users to be related to different Organisations for different Teams. In effect, users belong to Teams, which belong to Spaces, and in this way credit is able to cascade as required from a workflow to creators, contributors, submitters, the Teams and Spaces to which they belong, the consortia and organisations that these represent, and even new workflows which are derived from the original (see examples in **Figure 3**). This nested structure is capable of addressing sharing and credit for a diverse set of workflow contributors, including, but not limited to, individuals (e.g. workflow developers), research groups, institutions (e.g. universities), communities of practice (e.g. nf-core), and major research consortia (e.g. Biodiversity Genomics Europe (BGE), [https://biodiversitygenomics.eu/](https://biodiversitygenomics.eu/), [https://doi.org/10.3030/101059492](https://doi.org/10.3030/101059492)). For example, single users or research groups may only require a single Team to represent their workflow(s), and in this case they would add their Team to the default “Independent Teams” Space. However, a consortium representing multiple research groups or institutes may create a distinct Team for each one of its collaborating entities, and add the Teams to either the Independent Teams Space, or create a new Space to administer all these Teams. An individual, group, or consortium can thus establish a presence on WorkflowHub that reflects their real world structure, and which WorkflowHub then uses to assign credit. Multiple creators and Teams can be specified for a single workflow, additional credit can be assigned to contributors, and a distinction can be made between creators and submitters.

![figure3.png](figure3.png)

**Figure 3\.** A guide to the structures in WorkflowHub. You, the user, belong to one or more Organisations (i.e. affiliations). You can also belong to one or more Teams, each of which also needs to belong to a single Space (**top**). You can nominate which Organisations you wish to use for the different Teams that you have created or joined, and you can belong to multiple Teams in the same Space, as well as multiple Teams in other Spaces (**bottom**). This figure is also available in the WorkflowHub documentation [https://about.workflowhub.eu/docs/join-create-teams-spaces/](https://about.workflowhub.eu/docs/join-create-teams-spaces/).

The structure of WorkflowHub can also support an entire community of practice, whereby Spaces and Teams are used to organise contributors, share workflows, and support the sharing of knowledge within that community. This is achieved by registering workflows and linking them to other research outputs, including events, presentations, documents, publications, data files and SOPs \- these *assets* can either be hosted by WorkflowHub or added by reference. Workflows also support the nomination of relevant community channels, where developers can connect with users, and curated “Collections” of workflows can be constructed to help a community manage specific sets of relevant workflows, and other outputs, that may span many scientific applications, workflow languages and research programs (e.g. Threatened Species Initiative [https://threatenedspeciesinitiative.com/about-tsi/](https://threatenedspeciesinitiative.com/about-tsi/) annotation workflows [https://workflowhub.eu/collections/23](https://workflowhub.eu/collections/23), Australian BioCommons ‘Bring Your Own Data’ Expansion Project [https://workflowhub.eu/collections/6](https://workflowhub.eu/collections/6)). To streamline the onboarding process for consortia and larger projects, a generic set up guide \[[**Soiland-Reyes2024a**](https://doi.org/10.48546/WORKFLOWHUB.SOP.13.4)\], and multiple guides for specific consortia \[[**Soiland-Reyes2024b**](https://doi.org/10.48546/WORKFLOWHUB.SOP.14.1)\] \[[**Soiland-Reyes2024c**](https://doi.org/10.48546/WORKFLOWHUB.SOP.15.1)\] \[[**Goble2024a**](https://doi.org/10.48546/WORKFLOWHUB.SOP.10.1)\], have been created. These guides describe the steps required to get set up on WorkflowHub and use the platform effectively.

### A Hub for workflows

As the name of the registry suggests, the underlying aim of WorkflowHub is to act as a “Hub”, and specifically a Hub that helps to connect the workflows ecosystem and ease the interoperation of its constituent platforms and services. To realise this, WorkflowHub relies on a web-friendly metadata framework that simultaneously supports data representation within the registry itself, while also acting as a foundation for exchange of data and metadata between the platforms and services described in **Figure 1**. Here, we provide the four core examples that underpin this “Hub” functionality \[[**Goble2023**](https://doi.org/10.5281/zenodo.7886545)\]. 

1. **Authentication and Authorization Infrastructure:** WorkflowHub participates in the EOSC federated Authentication and Authorization Infrastructure through LS Login and the OAuth2 protocol, in addition to also supporting authentication via GitHub (see Section [WorkflowHub: a registry for computational workflows](https://docs.google.com/document/d/1pySWdXwMSLqznqDgMPvVN_LfA-DfHxikM-df__MxBkc/edit#heading=h.478murl1db09)). This feature allows systems interacting with WorkflowHub to identify users as the same individual across different systems – even with their identity provided by their participating institutional accounts – enabling single-sign on and smart authorization decisions on accessing and operating on workflows, their metadata and other resources across the ecosystem forming around WorkflowHub.  
2. **Workflow descriptions:** To help describe workflows and their components, the registry uses three profiles from bioschemas \[[**Gray2017**](https://iswc2017.semanticweb.org/paper-579/)\], including Computational [Tool](https://bioschemas.org/profiles/ComputationalTool/1.0-RELEASE), Computational [Workflow](https://bioschemas.org/profiles/ComputationalWorkflow/1.0-RELEASE) and Formal [Parameter](https://bioschemas.org/profiles/FormalParameter/1.0-RELEASE). Even though these are “bio” schemas, they are both domain and discipline independent, and support a standardised way of describing workflows and their components. In addition, and owing to close collaboration with the Common Workflow Language (CWL) \[[**Crusoe2022**](https://doi.org/10.1145/3486897)\]\[[**Amstutz2016**](https://doi.org/10.6084/m9.figshare.3115156.v2)\] community, the CWL workflow specification is encouraged as a workflow language independent way of describing a WorkflowHub entry. This is the so-called “Abstract CWL”. This format can even hold semantic annotations, which WorkflowHub leverages to extract the typing of workflow inputs and outputs, as well as EDAM ontology concepts for Topics and Operations \[[**Ison et al. 2013**](https://doi.org/10.1093/bioinformatics/btt113)\].   
3. **Interoperability:** RO-crate is a community-developed standard for FAIR Digital Objects \[[**De Smedt2020**](https://doi.org/10.3390/publications8020021)\] that is designed to package a workflow with the components required to understand and execute that workflow. The additional packaged components may include test data, Abstract CWL, diagrams, publications, and SOPs, as well as the flat file metadata that provides the context for all these [assets](https://www.researchobject.org/ro-crate/) \[[**Soiland-Reyes2022**](https://doi.org/10.3233/ds-210053)\]. The implementation of RO-crate is central to the ability of WorkflowHub to interoperate and exchange Workflow-RO-[Crate](https://about.workflowhub.eu/Workflow-RO-Crate/) data with the ecosystem in **Figure 1** (e.g. Zenodo archiving) and for exposing workflows as FAIR Digital Objects \[[**Soiland-Reyes2024d**](https://doi.org/10.5281/zenodo.13225792)\].   
4. **Execution:** WorkflowHub can interoperate with workflow execution platforms that are part of the ecosystem (see **Figure 1**) through its implementation of the GA4GH Tools Registry Service (TRS) API ([https://ga4gh.github.io/tool-registry-service-schemas/](https://ga4gh.github.io/tool-registry-service-schemas/), [https://www.ga4gh.org/product/tool-registry-service-trs/](https://www.ga4gh.org/product/tool-registry-service-trs/)). This means that a user of a TRS enabled platform, like Galaxy, is able to search for and retrieve workflows, without leaving the platform. It also means that WorkflowHub can send workflows to these platforms, ready for execution. 

Ultimately, the impact of these features is twofold: firstly WorkflowHub is able to support machine actionability, as described by the FAIR principles \[[**Wilkinson2016**\]](https://doi.org/10.1038/sdata.2016.18), and secondly the registry connects to services and platforms in use day-to-day by workflow users and developers.

### Using WorkflowHub to register, find and launch workflows

The primary purpose of WorkflowHub is to register and share workflows. Workflows can be viewed, downloaded and launched without registering with WorkflowHub. Open access and publicly accessible workflows are encouraged. However, workflows may also be registered privately, or be embargoed. This functionality supports creators of workflows in cases where they would like to (i.e. a workflow is still being developed), or need to, limit access to a specific group of users. The public content of WorkflowHub is openly accessible without login, while registering content with WorkflowHub requires user authentication to enable the assignment of ownership, credit, and citations.

To contribute content an individual user needs to:

1. **Register** and indicate the Organisations to which they are affiliated.  
2. **Decide which Space on WorkflowHub to use:** a Space is a user-administered section of WorkflowHub that can be used to manage the Teams required for consortia, institutes, or other large research activities. WorkflowHub administers a single default Space called “Independent Teams”. This Space can be used when users do not need to create and manage many Teams, but simply need to create a single Team. All other Spaces are created upon request and administered by those who requested the Space.  
3. **Create or join at least one Team:** every workflow in WorkflowHub is owned by at least one Team. WorkflowHub users must therefore belong to at least one Team, and this Team must belong to a Space (e.g. Independent Teams). Teams are one or more people working on a particular research activity involving workflows. In addition to enabling assignment of credit to workflow developers, the Team also enables its members to further describe the context for their work, and serves to promote their contributions to other WorkflowHub users. Note that workflows can be limited to single Teams or placed within multiple Teams, as needed.

Once these steps are complete, a contributor has the option to register:

* **Core resources** such as workflows and Collections. As workflows do not exist in a vacuum, Collections allow a WorkflowHub user to bring together workflows with any of their other resources and activities (see below) to create a visible and holistic resource that can support workflow reuse.  
* **Other resources**, including publications, documents, data files, and SOPs.  
* **Activities**, including presentations and events.

Registration of workflows can be carried out by manually uploading a workflow file, importing either a RO-crate or Git repository, and by submission through REST [API](https://about.workflowhub.eu/docs/adding-files/). RO-Crate is used by WorkflowHub as a fundamental unit that underpins upload, download, import and export. Importantly, a user does not need to know how to create or work with RO-Crates, as the registry automatically builds a crate when a workflow is registered. As a registry, WorkflowHub is designed to link to workflows held in their native repositories. However, because manual uploading and storage of files is also supported, it can also act as a repository.

To streamline these processes, each stage is guided by inbuilt wizards and users are prompted to carry out the next step. For example, after user registration with WorkflowHub, they are prompted to join or create a new Team. After registration of a workflow file or workflow repository, the user is prompted to complete the metadata suggested to create a well described and FAIR workflow entry.

**Figure 4** illustrates two example WorkflowHub entries: the *dna-seq-varlociraptor* Snakemake [workflow](https://workflowhub.eu/workflows/686) \[[**Köster2023**](https://doi.org/10.5281/zenodo.8421328)\], and the *Find transcripts \- TSI* Galaxy [workflow](https://workflowhub.eu/workflows/877) \[[**Silver2024**](https://doi.org/10.48546/WORKFLOWHUB.WORKFLOW.877.1)\]. The top of the entry emphasises the workflow name (**Figure 4B**) and workflow management system (**Figure 4A**), along with quick links to the development repository, requesting contact with the authors, subscribing to notifications about changes to the workflow, downloading a workflow research object crate (RO-Crate), and adding the workflow to a collection. The user can even download a RO-Crate that packages the files, references and metadata for a workflow. In addition, the workflow creator also has access to administrative options for the workflow entry in this section, which include adding new documents or presentations connected to the workflow, and workflow actions such as adding new versions, requesting a DOI for a specific workflow version, editing the workflow metadata, managing workflow contributors and visibility, and deleting the workflow entry (**Figure 4C**). The main panel for the entry has three tabs (**Figure 4D**) that provide a workflow overview (including descriptions, version information, metadata, critical annotations, and activity analytics), access and view capability for the files that were registered, and links to related items (e.g. people, Spaces, Teams, Collections, other workflows).

![figure4.png](figure4.png)

**Figure 4\.** Two example entries in WorkflowHub (**left:** \[[**Köster2023**](https://doi.org/10.5281/zenodo.8421328)\], **right:** \[[**Silver2024**](https://doi.org/10.48546/WORKFLOWHUB.WORKFLOW.877.1)\]) with sections of the user interface annotated and each entry using the flexible features of WorkflowHub in distinct ways. Entry features include **A**) workflow type, **B**) title, **C**) access panel with links to the source repository (e.g. GitHub), requests to contact the creators, subscribe / unsubscribe, download research object crate (RO-crate), add to a Collection, and in the right hand example access to administrative menus such as *Add new* (e.g. document, SOP) and *Actions* (e.g. edit or manage the workflow, including versions and minting DOIs), **D**) tabs for navigation between the entry overview, the list of files in the entry, and lists of items related to the workflow, including people, Teams, Spaces, Organisations, and other digital objects (e.g. publications, documents, SOPs, other workflows), **E**) description, which can be imported from Git, if available, **F**) version history, including Git commits, if available, **G**) creator and submitter information, **H**) links to more information about tools that comprise the workflow (i.e. bio.tools registry entries), **I**) license information, **J**) activity metrics (i.e. downloads and views), **K**) ontology concept annotations (e.g. EDAM in the left example entry), **L)** workflow diagram, **M**) parsed workflow inputs, outputs and steps for specific WfMS (e.g. Galaxy in the right example entry), **N**) buttons for launching workflows on execution platforms (e.g. Galaxy for right example entry), **O**) citation for the workflow (i.e. either using information from a minted DOI or a custom citation (e.g. workflow publication), **P**) custom tags, and **Q**) Collections that include the current workflow entry.

The examples use many of the key features of WorkflowHub, including those enabled by the WorkflowHub registration wizard, which prompts inclusion of critical metadata: these include title (**Figure 4B**), workflow management system (**Figure 4A**), creators (**Figure 4G**), component tools (**Figure 4H**), license information (**Figure 4I**) and ontology annotations (**Figure 4K**). In addition, the *dna-seq-varlociraptor* workflow made use of the WorkflowHub Git integration to ingest the repository README file (complete with badges), as well as providing the link to the development repository (**Figure 4C**), the ability to access and view the repository file list natively in WorkflowHub (**Figure 4D**), and an annotated version history (including commit IDs, **Figure 4F**). 

To find these workflows within the registry, a user can start by applying a text-based search, or by visiting the complete workflow listing where they can filter by type (e.g. Galaxy, Nextflow), tools used (i.e. using bio.tools identifiers \[[**Ison2019**](https://doi.org/10.1186/s13059-019-1772-6)\]), creators, Organisations, Teams, Spaces, and more. Researchers can also refine their searches by making use of faceted browsing and filtering on tags and other annotations, and are able to sort by titles, dates, views and downloads. 

Galaxy's integration with WorkflowHub leverages the GA4GH TRS API, enabling seamless workflow exchange. Researchers can discover, import, and run workflows from WorkflowHub directly within Galaxy. From the WorkflowHub interface users can utilise the "Run in Galaxy" button (see **Figure 4N**), which redirects them to a Galaxy instance and a workflow run form. This interoperability facilitates immediate application and further development of workflows. The use of RO-Crate specification ensures that workflow metadata and components remain accessible and interoperable, aligning with FAIR principles.

### Design that supports the workflow life cycle

To support the workflow life cycle \[[**Deelman2006**](https://doi.org/10.1109/E-SCIENCE.2006.261077)\]\[[**Courbebaisse2023**](https://doi.org/10.5281/zenodo.8324828)\], WorkflowHub integrates with services that workflow creators use for development, execution, maintenance, testing, citation, and ultimately archiving. These integrations initially form part of the workflow registration wizard, which guides a workflow creator through the process of registering their workflow for the first time. However, they are also accessible during a workflow’s maintenance phase, when the workflow may be updated to modify, improve, or repair its function.

#### Ease of access

A user of WorkflowHub begins by accessing the service via LS-Login. This enables researchers to use credentials from their specific organisations or even other identity-providing platforms (e.g. Google, Apple, ORCID). Authentication via GitHub or a local WorkflowHub account is also supported. This feature also means that WorkflowHub administrators can control and manage user access rights, and importantly can also create custom access levels that are suitable for specific user groups (e.g. research groups, consortia, international projects). For example, a contributor may simply want to register a single workflow to make it findable, and the contributor may be the only user that requires edit access to the workflow, or to the Team to which the workflow belongs, and access rights can be set accordingly. Conversely, a community of practice (e.g. nf-core), or consortium (e.g. BGE) may have multiple contributors, from multiple institutes, that also belong to multiple WorkflowHub Teams. In this case, granular permissions for edit rights can be set at the workflow, Team and Space levels.

#### Development and versioning 

Integration with the Git version control system is a key aspect of WorkflowHub that supports workflow creation, development and maintenance. If the workflow registration wizard is provided with a Git repository URL, WorkflowHub will automatically import and parse its metadata. In this case, a workflow creator only needs to review, and potentially update, metadata fields prior to completing the registration process. WorkflowHub’s Git integration supports workflows to remain in their native creation and development environment (i.e. a version control system), avoiding any impact of registration on the workflow’s development and management process. Moreover, it allows for automation of tasks like updating workflow entries in the registry when workflows are versioned (e.g., by using the LifeMonitor GitHub app [https://lifemonitor.eu/lm\_wft\_best\_practices\_github\_app](https://lifemonitor.eu/lm_wft_best_practices_github_app)) \[[**Goble2023**](https://doi.org/10.5281/zenodo.7886545)\] . For use cases like Galaxy, where workflows can be created via graphical user interface, rather than scripting, WorkflowHub provides the option to manually upload a workflow file, and step through the wizard manually to enter metadata.

#### WorkflowHub welcomes all workflows

Workflows come in all shapes and sizes, and may even be composed of multiple subworkflows. They may start off as a set of scripts and evolve into a heavily standardised and portable workflow. They may use one of the many known WfMS \[[**Amstutz2024**](https://s.apache.org/existing-workflow-systems)\]. And, of course, workflows can span virtually every field in the sciences and beyond. In short, the workflow and WfMS ecosystems are diverse. The role of WorkflowHub is to make these ecosystems transparent, and it does this by being agnostic to workflow language, maturity, source, structure, and even scientific quality. Contributions of every workflow type are encouraged (see [https://workflowhub.eu/workflow\_classes](https://workflowhub.eu/workflow_classes)); naturally, these metadata are presented to potential workflow consumers to readily identify workflows that may not be ready for reuse. Significant and popular WfMS may also be supported with relevant additional features, when appropriate (see use cases section).

#### Annotating workflow purpose (i.e. adding metadata)

In each case, the registration wizard guides users in provision of metadata. Metadata profiles for Computational Workflow and Formal Parameter (i.e. Bioschemas) are used in WorkflowHub, and ensure a discipline agnostic standard for describing workflows. Mandatory metadata fields are the workflow Title and the contributing Team(s). An indication of maturity of the workflow can be assigned by the creators. Workflows at any development stage (i.e. work-in-progress) are also encouraged, not just those workflows considered to be completed and stable. Workflow metadata can also be used to associate a workflow to relevant presentations, documents and data files.

Two key metadata integrations are also available. Without leaving WorkflowHub, workflows can be annotated with tool identifiers, either by automatic parsing of the workflow description, or by searching for and adding identifiers. This functionality is available for bio.tools software [identifiers](https://bio.tools/) and EDAM ontology concepts for both Topics (e.g. genomics) and data transformation Operations (e.g. genome assembly) \[[**Ison et al. 2013**](https://doi.org/10.1093/bioinformatics/btt113)\]. A creator can therefore annotate their workflow with persistent links to registry metadata about software that the workflow contains, and standardised shorthand terms that describe its application area and function. In the case of Galaxy, the standard structure of the workflow file is used to extract software tool identifiers and map these automatically to bio.tools \[[**Lamothe2023**](https://doi.org/10.37044/osf.io/79kje)\]. A user can build on these integrations by manually including custom tags and keywords. WorkflowHub also exposes metadata in a machine readable format (i.e. using Bioschemas), and in doing so makes this metadata accessible for indexing. This is in addition to the WorkflowHub API, which allows for retrieval and deposition of workflows. In each case, these different approaches assist with improving the findability of workflows and their components.

#### Discovery and understanding

When a workflow is ready to be shared and reused, a creator can update their workflow maturity from “work-in-progress” to “stable” in the WorkflowHub entry metadata. It is at this point that a workflow needs to be discoverable in multiple ways, and remain accessible in its original location. The reason for this is that not all researchers will necessarily seek to discover workflows in the same way. As a result, WorkflowHub is flexible in its approach. As mentioned earlier, within the registry itself a user can start by applying direct search and filtering to find workflows. External to the registry, the machine-readable, standardised format of WorkflowHub (i.e. Bioschemas) increases the search engine visibility of the rich metadata in a workflow entry. You can interactively explore the impact of workflow registration using evaluator tools such as FAIR-checker \[[**Rosnet2024**](https://github.com/IFB-ElixirFr/fair-checker)\].

Finding a workflow is step one for a potential user. Once found, the WorkflowHub entry metadata supports a user to understand the workflow, including its design, content, and purpose \[[**Cohen-Boulakia2017**](https://doi.org/10.1016/j.future.2017.01.012)\]. For example, “this is a Galaxy type workflow, containing these tools (i.e. links to bio.tools), which are capable of these types of data transformations (i.e. EDAM annotations)”. From a workflow entry that has been annotated with tools, a user can directly access the respective bio.tools entries to gain further understanding of the components that comprise the workflow. Users can also view the files in the source Git repository (if the workflow was imported from Git), and with a single click visit the repository. It is even possible to subscribe and be notified of changes to entries, removing the need for constant monitoring. 

#### Execution / reuse

WorkflowHub actively supports and develops integrations with workflow execution platforms and services. A key example is the GA4GH TRS [API](https://github.com/ga4gh/tool-registry-service-schemas). If an execution platform or system adopts the TRS standard, it can search WorkflowHub for suitable workflows, import those workflows, and execute them, without the need for the development of custom integrations with the workflow’s native repository. For example, Galaxy, Sapporo \[[**Suetake2022**](https://doi.org/10.12688/f1000research.122924.1)\] (DNA Data Bank of Japan (DDBJ, [https://ddbj.nig.ac.jp/](https://ddbj.nig.ac.jp/))), and the Workflow Execution Service (WfExS) ([https://github.com/inab/WfExS-backend](https://github.com/inab/WfExS-backend)) \[[**Fernandez2022**](https://doi.org/10.7490/f1000research.1119198.1)\] all implement the TRS API, either as providers or consumers. This allows these execution services to import and execute workflows from WorkflowHub. As execution platforms (e.g. Galaxy) can make use of the TRS, they are also able to provide inbuilt search interfaces that connect to WorkflowHub and support platform users to find and import a specific workflow. The connection of WorkflowHub to the LifeMonitor [service](https://www.lifemonitor.eu/), through the LifeMonitor GitHub app, allows workflow function and status to be reported to maintainers and users through regular automated tests driven by continuous integration (CI) based monitoring (e.g. Planemo automated workflow testing using Galaxy \[[**Bray2023**](https://genome.cshlp.org/content/33/2/261.long)\]). In these cases, WorkflowHub will also include a badge that shows if the tests are passing or failing. An embedded link in the badge takes a user to the LifeMonitor page for the workflow, providing information on the reliability of the workflow over time and the timeliness of the workflow maintainers in solving issues as they arise. In addition, the app can automatically suggest changes to the workflow Git repository that will bring its metadata and structure in line with best practices. Finally, to streamline the inclusion of reuse conditions, licensing information is included in the metadata for workflows registered with WorkflowHub, default sharing and license conditions can be specified for Teams, and these rules can be updated by Team administrators.

#### Attribution and citation

It is important to properly attribute contributions to workflows, and this includes provenance of the entire workflow development process. The RO-Crate format used by WorkflowHub allows for provenance tracking of metadata, ensuring that workflow creators are given the credit they deserve, while also adding accountability. Workflows can be linked to each other using WorkflowHub metadata (i.e. a workflow that is based on another workflow). The Git integrations described above also support the citation standard CITATION.[cff](https://citation-file-format.github.io/) \[[Druskat2021](https://doi.org/10.5281/zenodo.5171937)\], so that WorkflowHub can import this file and use its contents to populate workflow entry metadata for creators, including their [ORCID](https://orcid.org/). This approach simplifies citing a workflow according to the wishes of the workflow developer. Once a workflow is registered and credit is established within the metadata framework of WorkflowHub, the registry can also, at the push of a button, use [DataCite](https://datacite.org/) to mint persistent digital object identifiers (DOIs) for workflows and to contribute to the DataCite PID [Graph](https://support.datacite.org/docs/datacite-graphql-api-guide). This is the first step to ensuring that workflows can be cited effectively, increasing their visibility and potential impact, and supporting inclusion in the scholarly knowledge graph. 

### User engagement and training

WorkflowHub engages and supports a broad set of use cases, including numerous projects and consortia of global significance. Major projects are supported so that their members are able to use WorkflowHub in a way that aligns with the expectations of their project and its funders. Workflow communities (e.g. nf-core, CWL, Snakemake) are directly engaged by the WorkflowHub team, and supported to make their best practice workflows available in the registry.

The main mechanism through which engagement happens is the fortnightly open format WorkflowHub Club [meeting](https://about.workflowhub.eu/#community), where anyone can join the conversation, learn more about the registry, ask questions, and even contribute to the on-going development and evolution of the registry. There are also indirect ways through which workflow creators and users interact with WorkflowHub and its resources. The WorkflowHub Club team creates and presents content for the registry at conferences, in webinars and workshops, and as part of Ask-Me-Anything forum [events](https://about.workflowhub.eu/project/outreach/). Registered users of WorkflowHub are also able to ask questions and provide direct feedback via the registry interface. These communications are sent directly to the administrators of WorkflowHub for review and response. Finally, WorkflowHub operates a documentation site where users can access information on how to use the [registry](https://about.workflowhub.eu/docs/).

Clear and practical guidelines are required to support users of WorkflowHub. For example, some annotations are consistently missing from workflow entries, and useful features are sometimes overlooked (e.g. Git integration, parsing of CITATION.cff files, linking data to workflows, and linking workflows to each other). This highlights that the rich feature set of WorkflowHub is not necessarily immediately clear, and that guidance in leveraging these features is absolutely critical to support users in achieving best practice. It is important that the provided guidelines also extend to recommendations for how to organise a workflow Git repository. This will enable WorkflowHub to extend the process in place for Galaxy IWC to more community repositories: integrating with these repositories and automatically managing the update of workflow versions in WorkflowHub, such that they are in sync with Git releases. As a result, a general onboarding and set up guide for projects and consortia has been developed \[[**Soiland-Reyes2024a**](https://doi.org/10.48546/WORKFLOWHUB.SOP.13.4)\], as have multiple consortia specific guides \[[**Soiland-Reyes2024b**](https://doi.org/10.48546/WORKFLOWHUB.SOP.14.1)\]\[[**Soiland-Reyes2024c**](https://doi.org/10.48546/WORKFLOWHUB.SOP.15.1)\]\[[**Goble2024a**](https://doi.org/10.48546/WORKFLOWHUB.SOP.10.1)\]. Workflow resources have also been developed for the Galaxy Training Network (GTN \[[**Hiltemann2023**](https://doi.org/10.1371/journal.pcbi.1010752)\]) Smörgåsbord [events](https://gallantries.github.io/video-library/modules/ro-crate) and specific GTN tutorial [sets](https://training.galaxyproject.org/training-material/topics/fair/).

Finally, the WorkflowHub team actively identifies opportunities to engage with peer infrastructures to grow the user base of the registry, investigate and create integrations that are of enduring value, and further improve the function of the registry. For example, WorkflowHub is actively fostering a conversation with publishers and journals focused on how to make workflows citable objects in the literature \[[**Goble2024b**](https://galaxyproject.org/news/2024-08-03-workflow-publisher-forum/)\]. WorkflowHub is also being further developed to fully align with the best practice guidelines of the SciCodes Consortium \[[**Garijo2022**](https://doi.org/10.7717/peerj-cs.1023)\], implement the FAIR Principles for Research Software (FAIR4RS) \[[**Barker2022**](https://doi.org/10.1038/s41597-022-01710-x)\] and contribute to the CodeMeta specification for software \[[**Jones2024**](https://w3id.org/codemeta/v3.0)\]. As described earlier, a registry like WorkflowHub enables the creation of workflows that follow the FAIR principles, from the perspective of data \[[**Wilkinson2016**](https://doi.org/10.1038/sdata.2016.18)\], software \[[**Barker2022**](https://doi.org/10.1038/s41597-022-01710-x)\], and the unique features of workflows (e.g. abstraction, composition). WorkflowHub is central to discussions in the FAIR Computational Workflows working group for the Workflows Community Initiative (WCI, [https://workflows.community/groups/fair/](https://workflows.community/groups/fair/)). This working group engages broadly across the global workflows ecosystem (i.e. workflow developers, communities, platforms and services) to develop FAIR principles for workflows. The ultimate aim is to use the outcomes of these engagements to guide the evolution of WorkflowHub as a FAIR registry for workflows.

### Use Cases

To effectively support the sharing of workflows, WorkflowHub supports collaborations and communities of practice within the sciences. WorkflowHub contributions span domains such as cancer, COVID, genomics, rare diseases, geosciences, climate, physics, and more. WorkflowHub users span the globe and 35 countries are represented in the registered user list.

#### Research consortia & infrastructures

WorkflowHub is also an integral platform for consortia and projects. Here we provide details for three specific use cases, EOSC-Life, BGE, and Australian BioCommons.

EOSC-Life was a key use case driver, as it supported the implementation of FAIR computational workflows in the EU by seeking to develop a cloud-based Workflow Collaboratory \[[**Goble2021**](https://doi.org/10.5281/zenodo.4605654)\] that ultimately resulted in the creation of WorkflowHub. The aim was to support and streamline the sharing and reuse of FAIR computational workflows \[[**Goble2020**](https://doi.org/10.1162/dint_a_00033)\], in a way that is agnostic to research domain and research infrastructures boundaries, and that enables workflows to be convergence points for methods-based collaboration and reuse across these boundaries \[[**Goble2023**](https://doi.org/10.5281/zenodo.7886545)\]. WorkflowHub accommodates the diversity of EOSC-Life \[[**Goble2021**](https://doi.org/10.5281/zenodo.4605654)\] and ensures the visibility of workflows applied across its many established research infrastructures as they are created and registered.

The BGE project is a coming together of two communities of researchers with a common goal of cataloguing biodiversity through genomic resources: the European Reference Genome Atlas \[[**ERGA Mazzoni et al, 2023**](https://doi.org/10.1038/d41586-023-02229-w)\] and the European node of the International Barcode of Life consortium (iBOL Europe, [https://iboleurope.org/](https://iboleurope.org/)). Providing reference-quality genomes to the community (ERGA) and monitoring biodiversity through DNA barcoding (iBOL Europe) requires the management and processing of vast amounts of data, in an accessible and distributed fashion, relying on input from multiple individuals and institutes. The combination of BGE WorkflowHub Spaces (https://workflowhub.eu/programmes/25), Teams (https://workflowhub.eu/projects/163) and Collections (https://workflowhub.eu/collections/10) allows individuals to contribute as needed to the projects across the consortium. As workflows have been collected and curated by the community, they in effect also come with a “seal of approval” for external users that wish to replicate the work of ERGA or iBOL Europe. All together, such a structure of publishing and maintaining workflows facilitates BGE in achieving their ambitious goals of cataloguing biodiversity in Europe and bringing together researchers from the biodiversity genomics community.

Australian [BioCommons](https://www.biocommons.org.au/) is a national infrastructure project that actively supports life science research communities with community scale digital infrastructure. Rather than building anew, BioCommons aims to adopt fit-for-purpose international platforms and services that, in the case of workflows, can assist with the provision of sophisticated software, analysis capabilities, and digital asset stewardship. WorkflowHub is the primary workflow registry for [BioCommons](https://workflowhub.eu/programmes/8), and it is the focal point for the collaborative workflow efforts of BioCommons and its infrastructure partners together with Australian life science researchers.

#### Workflow management systems

WorkflowHub accepts all workflow types, and includes Galaxy \[[**GalaxyCommunity2022**](https://doi.org/10.1093/nar/gkac247)\], Snakemake \[[**Molder2021**](https://doi.org/10.12688/f1000research.29032.2)\], Nextflow \[[**DiTommaso2017**](https://doi.org/10.1038/nbt.3820)\]), job schedulers (e.g. PyCOMPS \[[**Tejedor2017**](https://doi.org/10.1177/1094342015594678)\]), application-specific types like SCIPION \[[**De La Rosa-Trevín2016**](https://doi.org/10.1016/j.jsb.2016.04.010)\], notebooks (e.g. Jupyter, [https://jupyter.org/](https://jupyter.org/)), and even scripting languages (e.g. R \[[**R Core Team 2024**](https://www.r-project.org/)\] and Python \[[**Python 2024**](https://www.python.org/)\]) (see also **Figure 2**). WorkflowHub provides customised support for WfMS that are critical for specific domain communities (e.g. bioinformatics). This is currently the case for Galaxy, CWL and Nextflow. Galaxy is the most registered workflow type. The registry development team, together with Galaxy, have created functionalities for WorkflowHub that include 1\) semi-automated registration of new and updated workflows versions from the Galaxy IWC, 2\) integration with LifeMonitor to further support semi-automated registration but also Planemo test monitoring, and 3\) automatic mapping of tool identifiers to the bio.tools registry. As the CWL community has been closely involved in registry development, WorkflowHub also has in-built functions for parsing CWL. Most recently, engagement between nf-core and the WorkflowHub team has resulted in automatic registration and metadata parsing for this community's Nextflow [workflows](https://elixiruknode.org/news/2024/workflowhub-nf-core-workflow-accessibility/).

#### Individual workflow developers

WorkflowHub supports large international science missions and consortia. However, the registry welcomes and encourages contributions from any workflow developer, regardless of their research domain, application areas, or the size of their research group. In fact, the largest Space on WorkflowHub is currently Independent Teams, which covers 287 people, 185 Teams, 180 Organisations and 279 workflows.

## Discussion

We have presented WorkflowHub, a registry that enriches the scientific workflows ecosystem by being a hub for discovery and sharing of workflows from across multiple languages, communities, consortia, and scientific domains. The registry connects this community of users and contributors to workflow development, support, and scholarly services that support the requirement for workflows to be shared and credited, as well as the various requirements developers encounter during the workflow life cycle (see **Figure 1**). Workflows continually evolve to keep pace with changing research questions, data types and practices. Similarly, it is intended for WorkflowHub to evolve in lock step with the changing requirements of both the developers and users of computational workflows. The roadmap for WorkflowHub over the next few years can be broadly broken down into four aspects: improving the support and resources available to users of the registry, onboarding new communities and domains, contributing thought leadership for workflow best practices that directly impact aspects such as workflow visibility and quality, and aligning to FAIR principles for computational workflows.

### Improving support

As WorkflowHub and its integrated services intend to support the living lifecycle of a workflow, and not simply the registration and publication of workflows, the registry will aim to improve processes that align to supporting this life cycle. This includes improving search and the inbuilt wizards that assist with onboarding and resource registration, but also improving overall ease-of-use and guidance. Additions and improvements for the UI will continue, adding those features that are required by end users. Particular attention will be paid to tracking workflow cloning, as well as supporting workflow collections, sub-workflows, and nested workflows. 

As the number of contributors and workflow assets grow, it will be critical to explore the use of automated mechanisms that allow scaling of support. This could include streamlined metadata annotation approaches that use Large Language Models (LLMs) to populate metadata fields for review by a workflow creator, automatic reporting of issues to workflow creators and their Teams, or integration with additional added value services like APICURON \[[**Hatos2021**](https://doi.org/10.1093/database/baab019)\]. 

Continued work on the onboarding guidelines created for WorkflowHub will improve the quality of workflow organisation, for both contributing groups (i.e. big projects and consortia) and the workflows themselves. Additional guidance will include documenting best practices for WfMS (e.g. repository layout guidelines). To date, WorkflowHub has been included in existing training, such as the GTN Smörgå[sbords](https://training.galaxyproject.org/). As a next step, this effort should be extended to include WorkflowHub lessons in Software Carpentry developed by the Data Science training programme for Health and Biosciences ([Ed-DASH](https://edcarp.github.io/Ed-DaSH/)) for [Nextflow](https://carpentries-incubator.github.io/workflows-nextflow/) and [Snakemake](https://carpentries-incubator.github.io/snakemake-novice-bioinformatics/). 

Reflecting their importance to the operation of WorkflowHub, integrations with platforms and services that support the workflow life cycle will be added, improved, and updated. The RO-Crate format used by WorkflowHub will be central to this effort, as it enables exchange of workflows with other services. Planned updates include the ability to natively configure automated synchronisation between Git repositories and WorkflowHub. Setting this up to match the expectations of the community of WorkflowHub users will be essential, and is likely to involve collaborative work with developers that operate the repositories used by workflow communities. Clear guidance on the best way to leverage the LifeMonitor GitHub App will also be added, as will the ability to “launch” workflows on additional instances that implement TRS (i.e. Seqera Platform [https://seqera.io/platform/](https://seqera.io/platform/)). An integration to link from WorkflowHub entries to workflow benchmarking information in [OpenEBench](https://openebench.bsc.es/) is also planned. \[[**Capella-Gutierrez2017**](https://doi.org/10.1101/181677)\].

### Onboarding

WorkflowHub already has semi-automated support in place for WfMS like Galaxy, Nextflow and CWL. WorkflowHub recently extended these to onboard the entire effort of the nf-core community of practice and are finalising a similar integration for the GTN. These specific integrations automatically add hundreds of workflows to the registry, with support in place to auto-register new workflows, and workflow versions, as they are released by these communities. Identifying and engaging workflow communities to onboard, in particular those that converge on specific WfMS, will be critical to increasing the number of workflows in the registry and the visibility of the workflows ecosystem overall, as well as to ensure that further requirements for WorkflowHub are being collected from a diverse set of stakeholders. Support provided to new communities may entail tailored integration and semi-automated ingestion of workflows from community repositories, as well as synchronised registration of new workflow releases and supporting WfMS communities to make use of both RO-Crate and Abstract CWL. Such support is now being explored for Snakemake workflows, and this will require close collaboration with the Snakemake community, including developers.

### Workflow visibility and recognition

WorkflowHub already has the capacity to harvest citation, author and contributor credit metadata from CITATION.cff files. In future, this could be extended to other popular standards, including codemeta.json \[[**Jones2024**](https://w3id.org/codemeta/v3.0)\]. Through DataCite, WorkflowHub has the means to mint DOIs, and incorporate workflows with DOIs into [OpenAIRE](https://www.openaire.eu/). WorkflowHub has thus actively worked to increase the visibility of workflows in a standardised and streamlined fashion. This approach is already bearing fruit, with examples of WorkflowHub formatted citations appearing in the published literature \[[**Hall2024**](https://doi.org/10.1093/gigascience/giae010)\]\[[**Roach2024**](https://doi.org/10.1093/gigascience/giae020)\]. In addition, there are now examples of journals making workflows the focus of published works. A critical outcome here is to ensure that WorkflowHub becomes a recommended registry for journals and publishers. This includes providing workflow creators and users with a set of best practice recommendations for how to properly document and cite workflows in published research. A forum has been established already with multiple publishers and journals, and these continuing conversations will aim to address the challenges (i.e. citation formats, complexity of workflow citations, impact on publishing system, recommended practices for peer review of workflows) and opportunities (i.e. developer recognition, tracking, reproducibility) presented by formal citation of computational workflows.

### Hand-holding for FAIR principles

As frequently alluded to in other sections, WorkflowHub is strongly connected to the FAIR principles at numerous levels. One of these levels that is of particular importance for users is the way that it “holds users’ hands” during the process of sharing workflows and related research artefacts. The FAIR principles are only guidelines, and even the most well-intentioned attempts to follow them can go awry in unexpected ways \[[**Wilkinson2022**](https://doi.org/10.1109/eScience55777.2022.00090)\]. WorkflowHub aims to help the user to follow best practices, including following the FAIR principles, by making them convenient, but not imposing them as requirements.

For example, WorkflowHub helps users make their workflows Findable – easy to find for both humans and machines – by assigning globally unique and persistent identifiers for the workflows and their different versions (i.e. WorkflowHub identifiers and DOIs). WorkflowHub also guides users to describe their workflows with rich metadata, including the identifier for the workflow, and these metadata are automatically exposed for indexing through WorkflowHub’s use of Bioschemas.

WorkflowHub similarly helps users make their workflows Accessible – available to humans and machines over open protocols that provide optional access control – by providing multiple APIs over HTTPS. These APIs include the JSON-based FAIRDOM-SEEK [API](https://workflowhub.eu/api), an RO-Crate Submission [API](https://about.workflowhub.eu/developer/ro-crate-api/), and a Tool Registry Service API from the open and free GA4GH TRS standard. Workflow DOIs can also be minted, which means the workflow and its metadata will be accessible even if the workflow itself is no longer available or if the workflow itself cannot be shared openly.

Workflows, after being found and accessed, should ideally be Interoperable – able to be used by humans and machines as part of a wider computational ecosystem. Interoperability often requires a lot of “plumbing” that WorkflowHub provides for users automatically through the use of open-source standards (i.e. RO-Crate as the primary data exchange format) and domain- and tool-specific integrations (i.e. Galaxy IWC and Nextflow nf-core). By guiding users through the process of inputting metadata, WorkflowHub reduces complexity and tedium, making it significantly easier to create interoperable workflows.

Finally, WorkflowHub helps users ensure their workflows are Reusable – allowed to be used in part or in entirety by other humans and machines. In particular, it does this by providing opportunities to specify a clear and accessible license, qualified references to other software, and detailed provenance. WorkflowHub also collaborates closely with prominent communities in the computational workflows space (see Use Cases section) so that the registry can accommodate and incorporate domain-relevant community standards.

## Methods

### Governance

WorkflowHub is an ELIXIR service supported by the UK and Belgium ELIXIR [Nodes](https://elixir-europe.org/) \[[**ELIXIR2024**](https://doi.org/10.7490/f1000research.1119751.1)\]) as well as Australian [BioCommons](https://www.biocommons.org.au/) \[[**Francis2024**](https://doi.org/10.5281/zenodo.13626350)\]. The registry is part of both the ELIXIR Tools [Platform](https://elixir-europe.org/platforms/tools) and Research Software [Ecosystem](https://research-software-ecosystem.github.io/), and forms part of both [EuroScienceGateway](https://esciencelab.org.uk/projects/eurosciencegateway/) and DARE-UK TRE-[FX](https://esciencelab.org.uk/projects/tre-fx/). Governance is coordinated and managed by the WorkflowHub [Club](https://about.workflowhub.eu/project/community/): an inclusive community that meets biweekly online and consists of workflow developers/creators and workflow users, as well as WorkflowHub developers and product owners. The minutes for these meetings are open \[[link](https://docs.google.com/document/d/1U2KAlbKviCu-fCX-znncKIBUIUUOeEnuRGdAg-fNd4Q/edit)\] and a GitHub organisation is used to manage [documentation](https://github.com/workflowhub-eu/about). Club members include representatives from ELIXIR, WCI, Australian BioCommons, and more. Over 60 people are listed as [contributors](https://about.workflowhub.eu/project/acknowledgements/#workflowhub-club), and any new contributors are invited to join. WorkflowHub has secured funding for sustainability through Horizon Europe projects, ELIXIR Europe and national UK funds. 

### Technical infrastructure

WorkflowHub is developed openly, and largely virtually, using open software development practices, hackathons, and virtual communication channels. It has both a [roadmap](https://about.workflowhub.eu/project/roadmap/) and regular release cycle (i.e SEEK release cycle, [https://github.com/seek4science/seek/](https://github.com/seek4science/seek/)). WorkflowHub requests for Team registration are supported for the American and Asia Pacific time zones by Oak Ridge National Laboratory and Australian BioCommons, respectively. WorkflowHub is built on the FAIRDOM-SEEK software \[[**Wolstencroft2017**](https://doi.org/10.1093/nar/gkw1032)\]. FAIRDOM-SEEK was originally developed as a data-management platform for the systems biology community, but has been generalised over time to support a wide variety of use cases, and now has numerous deployments across the world supporting many different communities. Development of FAIRDOM-SEEK is a collaborative activity with contributors from institutions in the UK, Germany, Belgium, Sweden and elsewhere. WorkflowHub is currently hosted on the University of Manchester's Research IT cloud. WorkflowHub makes use of Git to store workflows as repositories. This enables the addition, modification and deletion of files for workflows that are uploaded directly to the registry by a user, as well as the ability to freeze “snapshots” of specific workflow versions. If a workflow has been added either via Git import or through submission of an RO-Crate, it then needs to be explicitly versioned and uploaded again. Curation of registered workflows is the responsibility of the workflow creators and / or submitters.

## Data Availability

WorkflowHub will ensure data and metadata availability and access up to 2027, with plans to extend this availability further. Further availability of data and metadata from WorkflowHub falls under two categories. For workflows with minted DOIs, the archiving of data and metadata is managed via DataCite. For all workflows, and associated digital assets, WorkflowHub has an End-of-Life policy. The policy states that *“If and when the WorkflowHub reaches its end of service after that (i.e. 2027), the published contributions and metadata will be archived as RO-Crates and made available through a public repository, such as Zenodo, Figshare or another appropriate resource at that time. DOI registrations will in this case be updated to link to the archived deposits.”* (see [https://about.workflowhub.eu/project/\#retention-and-end-of-life-policy](https://about.workflowhub.eu/project/#retention-and-end-of-life-policy)). A knowledge graph of registered Workflow RO-Crates as of 2024-08 is published on Zenodo \[[**Hambley2024**](https://doi.org/10.5281/zenodo.13362051)\].

## Code Availability

WorkflowHub code is available as part of the FAIRDOM-SEEK project ([https://github.com/seek4science/seek](https://github.com/seek4science/seek) \[[**Owen2024**](https://doi.org/10.5281/zenodo.5653415)\]).

## Acknowledgments

The authors are grateful to Sehrish Kanwal of the University of Melbourne for feedback and suggestions about the draft.

## Funding

workflowhub.eu was founded as part of EOSC-Life (WP2 Tools Collaboratory), funded by the European Union’s Horizon 2020 programme under grant agreement H2020-INFRAEOSC-2018-2 824087\. This work was further supported by funding from European Commission’s Horizon Europe programme and UK Research and Innovation (UKRI) under the UK government’s Horizon Europe funding guarantee: EuroScienceGateway (HORIZON-INFRA-2021-EOSC-01-04 101057388, UKRI 10038963), BY-COVID (HORIZON-INFRA-2021-EMERGENCY-01 101046203), FAIR-IMPACT (HORIZON-INFRA-2021-EOSC-01-05 101057344, UKRI 10038992), BioDT (HORIZON-INFRA-2021-TECH-01-01 101057437, UKRI 10038930), PREP-IBISBA (H2020-INFRADEV-2019-2 871118), AgroServ (HORIZON-INFRA-2021-SERV-01-02 101058020, UKRI 10038927), BIOINDUSTRY 4.0 (HORIZON-INFRA-2022-TECH-01 101094287, UKRI 10048146). 

This work is supported by Australian BioCommons which is enabled by NCRIS via Bioplatforms Australia funding, by the Sardinian Regional Government through the XData Project, and by the LIFEMAP project (Italian Ministry of Health, POS T3).

The work supported by the ELIXIR-DE node was supported by the German Federal Ministry of Education and Research BMBF grant 031 A538A de.NBI-RBC and the Ministry of Science, Research and the Arts Baden-Württemberg (MWK) within the framework of LIBIS/de.NBI Freiburg.

This research used resources of the Oak Ridge Leadership Computing Facility at the Oak Ridge National Laboratory, which is supported by the Office of Science of the U.S. Department of Energy under Contract No. DE-AC05-00OR22725.

## Author Contributions

[https://credit.niso.org](https://credit.niso.org/): Conceptualization, Data curation, Formal analysis, Funding acquisition, Investigation, Methodology, Project administration, Software, Resources, Supervision, Validation, Visualization, Writing – original draft, Writing – review & editing

| Author | Contributions |
| :---: | :---- |
| OJRG | Writing – original draft, Writing – review & editing, Project administration, Data curation, Supervision |
| SW | Writing – original draft, Writing – review & editing, Project administration |
| FB | Conceptualization, Data curation, Investigation, Project administration, Software, Methodology, Resources |
| SL | Writing \- review & editing |
| LP | Writing \- review & editing |
| SSR | Conceptualization, Funding acquisition, Investigation, Supervision, Visualization, Writing – review & editing |
| SO | Conceptualization, Investigation, Project administration, Software, Methodology  |
| NJ |  |
| FC |  |
| CG | Conceptualization, Funding acquisition, Methodology, Project administration, Resources, Supervision, Visualization, Writing – original draft, Writing – review & editing |
| BG | Writing \- review & editing |
| JMF | Writing \- review & editing |
| HM | Writing \- review & editing |
| SCG | Writing \- review & editing |
| TB | Writing \- review & editing |

## References

1. Goble, C. WorkflowHub Publishers and Journal Forum. https://galaxyproject.org/news/2024-08-03-workflow-publisher-forum/ (2024).  
2. Francis, R. & Christiansen, J. H. Australian BioCommons Strategic Plan 2023 \- 2028\. Preprint at https://doi.org/10.5281/ZENODO.13626350 (2024).  
3. Wolstencroft, K. et al. FAIRDOMHub: a repository and collaboration environment for sharing systems biology research. Nucleic Acids Res 45, D404–D407 (2017).  
4. Capella-Gutierrez, S. et al. Lessons Learned: Recommendations for Establishing Critical Periodic Scientific Benchmarking. Preprint at https://doi.org/10.1101/181677 (2017).  
5. Goble, C. et al. EOSC-Life Implementation of a mechanism for publishing and sharing workflows across instances of the environment. (2023) doi:10.5281/ZENODO.7886545.  
6. Patel, R. et al. Reproducibility of the First Image of a Black Hole in the Galaxy M87 from the Event Horizon Telescope (EHT) Collaboration. (2022) doi:10.48550/ARXIV.2205.10267.  
7. De Smedt, K., Koureas, D. & Wittenburg, P. FAIR Digital Objects for Science: From Data Pieces to Actionable Knowledge Units. Publications 8, 21 (2020).  
8. Owen, S. et al. Seek4science/Seek: FAIRDOM-SEEK v1.15.0. (Zenodo, 2024). doi:10.5281/zenodo.11209855.  
9. Freudling, W. et al. Adaptive Data Reduction Workflows for Astronomy \-- The ESO Data Processing System (EDPS). (2023) doi:10.48550/ARXIV.2311.03822.  
10. McClure, J. E. et al. Toward Real-Time Analysis of Synchrotron Micro-Tomography Data: Accelerating Experimental Workflows with AI and HPC. in Driving Scientific and Engineering Discoveries Through the Convergence of HPC, Big Data and AI (eds. Nichols, J. et al.) vol. 1315 226–239 (Springer International Publishing, Cham, 2020).  
11. Wratten, L., Wilm, A. & Göke, J. Reproducible, scalable, and shareable analysis pipelines with bioinformatics workflow managers. Nat Methods 18, 1161–1168 (2021).  
12. Reiter, T. et al. Streamlining data-intensive biology with workflow systems. GigaScience 10, giaa140 (2021).  
13. The ICGC/TCGA Pan-Cancer Analysis of Whole Genomes Consortium et al. Pan-cancer analysis of whole genomes. Nature 578, 82–93 (2020).  
14. Cohen-Boulakia, S. et al. Scientific workflows for computational reproducibility in the life sciences: Status, challenges and opportunities. Future Generation Computer Systems 75, 284–298 (2017).  
15. Amstutz, P., Mikheev, M., Crusoe, M. R., Tijanić, N. & Lampa, S. Existing Workflow systems. Common Workflow Language wiki. GitHub https://s.apache.org/existing-workflow-systems.  
16. Existing Workflow systems. GitHub https://github.com/common-workflow-language/common-workflow-language/wiki/Existing-Workflow-systems.  
17. Hiltemann, S. et al. Galaxy Training: A powerful framework for teaching\! PLoS Comput Biol 19, e1010752 (2023).  
18. Bray, S. et al. The Planemo toolkit for developing, deploying, and executing scientific data analyses in Galaxy and beyond. Genome Res. 33, 261–268 (2023).  
19. Ison, J. et al. The bio.tools registry of software tools and data resources for the life sciences. Genome Biol 20, 164 (2019).  
20. Silver, L. & Syme, A. Find transcripts \- TSI. (2024) doi:10.48546/WORKFLOWHUB.WORKFLOW.877.1.  
21. Goble, C. et al. EOSC-Life Implementation of a mechanism for publishing and sharing workflows across instances of the environment. (2023) doi:10.5281/ZENODO.7886545.  
22. Wilkinson, S. R. et al. F\*\*\* workflows: when parts of FAIR are missing. in 2022 IEEE 18th International Conference on e-Science (e-Science) 507–512 (IEEE, Salt Lake City, UT, USA, 2022). doi:10.1109/eScience55777.2022.00090.  
23. Hatos, A., Quaglia, F., Piovesan, D. & Tosatto, S. C. E. APICURON: a database to credit and acknowledge the work of biocurators. Database 2021, baab019 (2021).  
24. ELIXIR. ELIXIR Annual Report 2023\. (2024) doi:10.7490/F1000RESEARCH.1119751.1.  
25. Lonie & Francis. Australian BioCommons Strategic Plan  2019-2023. (2020).  
26. Jones, M. B. et al. CodeMeta: An Exchange Schema for Software Metadata. (2024).  
27. Druskat, S. et al. Citation File Format. (2021) doi:10.5281/ZENODO.5171937.  
28. Rosnet, T., Gaignard, A., Devignes, M.-D. & Frikha, S. FAIR-Checker. (2024).  
29. Amstutz, P. et al. Common Workflow Language, v1.0. Preprint at https://doi.org/10.6084/M9.FIGSHARE.3115156.V2 (2016).  
30. Crusoe, M. R. et al. Methods included: standardizing computational reuse and portability with the Common Workflow Language. Commun. ACM 65, 54–63 (2022).  
31. Gray, A. Bioschemas: From Potato Salad to Protein Annotation. (2017).  
32. Köster, J. et al. Snakemake-Workflows/Dna-Seq-Varlociraptor: V5.0.2. (\[object Object\], 2023). doi:10.5281/ZENODO.8421328.  
33. Soiland-Reyes, S. et al. Packaging research artefacts with RO-Crate. DS 5, 97–138 (2022).  
34. Stevens, F. Understanding How Researchers Find Research Software for Research Practice. https://zenodo.org/record/7340034 (2022) doi:10.5281/ZENODO.7340034.  
35. Courbebaisse, G. et al. Research Software Lifecycle. (2023) doi:10.5281/ZENODO.8324828.  
36. Welter, D. et al. FAIR in action \- a flexible framework to guide FAIRification. Sci Data 10, 291 (2023).  
37. Barker, M. et al. Introducing the FAIR Principles for research software. Sci Data 9, 622 (2022).  
38. Wilkinson, M. D. et al. The FAIR Guiding Principles for scientific data management and stewardship. Sci Data 3, 160018 (2016).  
39. Garijo, D. et al. Nine best practices for research software registries and repositories. PeerJ Computer Science 8, e1023 (2022).  
40. Roach, M. J. et al. Hecatomb: an integrated software platform for viral metagenomics. GigaScience 13, giae020 (2024).  
41. Hall, M. B. & Coin, L. J. M. Pangenome databases improve host removal and mycobacteria classification from clinical metagenomic data. GigaScience 13, giae010 (2024).  
42. Soiland-Reyes, S. The BGE guide to using WorkflowHub. Preprint at https://workflowhub.eu/sops/15?version=1 (2024).  
43. Soiland-Reyes, S. BioDT Guide to using WorkflowHub. Preprint at https://workflowhub.eu/sops/14?version=1 (2024).  
44. Deelman, E. & Gil, Y. Managing Large-Scale Scientific Workflows in Distributed Environments: Experiences and Challenges. in 2006 Second IEEE International Conference on e-Science and Grid Computing (e-Science’06) 144–144 (2006). doi:10.1109/E-SCIENCE.2006.261077.  
45. Mazzoni, C. J., Ciofi, C. & Waterhouse, R. M. Biodiversity: an atlas of European reference genomes. Nature 619, 252–252 (2023).  
46. Python. (2024).  
47. R Core Team. R: A Language and Environment for Statistical Computing.  
48. De La Rosa-Trevín, J. M. et al. Scipion: A software framework toward integration, reproducibility and validation in 3D electron microscopy. Journal of Structural Biology 195, 93–99 (2016).  
49. Goble, C., Bacall, F. & Soiland-Reyes, S. The BY-COVID Guide to using WorkflowHub. Preprint at https://workflowhub.eu/sops/10?version=1 (2024).  
50. Soiland-Reyes, S., Goble, C., Bacall, F., Gustafsson, J. & Andrade Buono, R. A guide to using WorkflowHub. Preprint at https://workflowhub.eu/sops/13?version=4 (2024).  
51. Fernández, J. M., Rodríguez-Navas, L. & Capella-Gutiérrez, S. Secured and annotated execution of workflows with WfExS-backend. (2022) doi:10.7490/F1000RESEARCH.1119198.1.  
52. Suetake, H. et al. Sapporo: A workflow execution service that encourages the reuse of workflows in various languages in bioinformatics. F1000Res 11, 889 (2022).  
53. The Snakemake team. Snakemake workflow catalog. https://snakemake.github.io/snakemake-workflow-catalog/.  
54. Mölder, F. et al. Sustainable data analysis with Snakemake. F1000Res 10, 33 (2021).  
55. Di Tommaso, P. et al. Nextflow enables reproducible computational workflows. Nat Biotechnol 35, 316–319 (2017).  
56. The Galaxy Community et al. The Galaxy platform for accessible, reproducible, and collaborative data analyses: 2024 update. Nucleic Acids Research gkae410 (2024) doi:10.1093/nar/gkae410.  
57. Ison, J. et al. EDAM: an ontology of bioinformatics operations, types of data and identifiers, topics and formats. Bioinformatics 29, 1325–1332 (2013).  
58. Ison, J. et al. Tools and data services registry: a community effort to document bioinformatics resources. Nucleic Acids Res 44, D38–D47 (2016).  
59. Baker, D. et al. No more business as usual: Agile and effective responses to emerging pathogen threats require open data and open analytics. PLoS Pathog 16, e1008643 (2020).  
60. Maier, W. et al. Ready-to-use public infrastructure for global SARS-CoV-2 monitoring. Nat Biotechnol 39, 1178–1179 (2021).  
61. Larivière, D. et al. Scalable, accessible and reproducible reference genome assembly and evaluation in Galaxy. Nat Biotechnol 42, 367–370 (2024).  
62. Deelman, E. et al. The future of scientific workflows. The International Journal of High Performance Computing Applications 32, 159–175 (2018).  
63. Gil, Y. et al. Examining the Challenges of Scientific Workflows. Computer 40, 24–32 (2007).  
64. Deelman, E. & Chervenak, A. Data Management Challenges of Data-Intensive Scientific Workflows. in 2008 Eighth IEEE International Symposium on Cluster Computing and the Grid (CCGRID) 687–692 (2008). doi:10.1109/CCGRID.2008.24.  
65. Australian BioCommons. Australian BioCommons https://www.biocommons.org.au (2024).  
66. Lamothe, L. et al. An Evaluation of EDAM Coverage in the Tools Ecosystem and Prototype Integration of Galaxy and WorkflowHub Systems. https://osf.io/79kje (2023) doi:10.37044/osf.io/79kje.  
67. Goble, C. et al. WorkflowHub – a FAIR registry for workflows. (2022) doi:10.7490/F1000RESEARCH.1118984.1.  
68. Yuen, D. et al. The Dockstore: enhancing a community platform for sharing reproducible and accessible computational protocols. Nucleic Acids Research 49, W624–W632 (2021).  
69. Ewels, P. A. et al. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol 38, 276–278 (2020)  
70. Common Workflow Language Viewer. https://view.commonwl.org/workflows.  
71. Tejedor, E. et al. PyCOMPSs: Parallel computational workflows in Python. The International Journal of High Performance Computing Applications 31, 66–82 (2017).  
72. The Galaxy Community et al. The Galaxy platform for accessible, reproducible and collaborative biomedical analyses: 2022 update. Nucleic Acids Research 50, W345–W351 (2022).  
73. Goble, C. et al. FAIR Computational Workflows. Data Intellegence 2, 108–121 (2020).  
74. Goble, C. et al. Implementing FAIR Digital Objects in the EOSC-Life Workflow Collaboratory. (2021) doi:10.5281/ZENODO.4605654.
